spring-integration-file/src/main/java/org/springframework/integration/file/FileWritingMessageHandler.java

```diff
@@ -608,4 +730,216 @@ else if (destinationDirectoryToUse instanceof File) {
+		flushIfNeeded(this.flushPredicate, message);
+	}
+
+	/**
+	 * When using {@link FileExistsMode#APPEND_NO_FLUSH} you can invoke this method to
+	 * selectively flush open files. For each open file the supplied
+	 * {@link MessageFlushPredicate#shouldFlush(String, long, Message)}
+	 * method is invoked and if true is returned, the file is flushed.
+	 * @param flushPredicate the {@link FlushPredicate}.
+	 * @since 4.3
+	 */
+	public void flushIfNeeded(FlushPredicate flushPredicate) {
+		Iterator<Entry<String, FileState>> iterator = FileWritingMessageHandler.this.fileStates.entrySet().iterator();
+		while (iterator.hasNext()) {
+			Entry<String, FileState> entry = iterator.next();
+			FileState state = entry.getValue();
+			if (flushPredicate.shouldFlush(entry.getKey(), state.lastWrite)) {
+				iterator.remove();
+				state.close();
+			}
+		}
+	}
+
+	/**
+	 * When using {@link FileExistsMode#APPEND_NO_FLUSH} you can invoke this method to
+	 * selectively flush open files. For each open file the supplied
+	 * {@link MessageFlushPredicate#shouldFlush(String, long, Message)}
+	 * method is invoked and if true is returned, the file is flushed.
+	 * @param flushPredicate the {@link MessageFlushPredicate}.
+	 * @param filterMessage an optional message passed into the predicate.
+	 * @since 4.3
+	 */
+	public void flushIfNeeded(MessageFlushPredicate flushPredicate, Message<?> filterMessage) {
+		Iterator<Entry<String, FileState>> iterator = FileWritingMessageHandler.this.fileStates.entrySet().iterator();
+		while (iterator.hasNext()) {
+			Entry<String, FileState> entry = iterator.next();
+			FileState state = entry.getValue();
+			if (flushPredicate.shouldFlush(entry.getKey(), state.lastWrite, filterMessage)) {
+				iterator.remove();
+				state.close();
+			}
+		}
+	}
+
+
+	private static final class FileState {
+
+		private final BufferedWriter writer;
+
+		private final BufferedOutputStream  stream;
+
+		private volatile long lastWrite;
+
+		private FileState(BufferedWriter writer) {
+			this.writer = writer;
+			this.stream = null;
+		}
+
+		private FileState(BufferedOutputStream stream) {
+			this.writer = null;
+			this.stream = stream;
+		}
+
+		private void close() {
+			try {
+				if (this.writer != null) {
+					this.writer.close();
+				}
+				else {
+					this.stream.close();
+				}
+			}
+			catch (IOException e) {
+				;
+			}
+		}
+	}
+
+	private final class Flusher implements Runnable {
+
+		@Override
+		public void run() {
+			synchronized (FileWritingMessageHandler.this) {
+				long expired = FileWritingMessageHandler.this.flushTask == null ? Long.MAX_VALUE
+						: (System.currentTimeMillis() - FileWritingMessageHandler.this.flushInterval);
+				Iterator<Entry<String, FileState>> iterator = FileWritingMessageHandler.this.fileStates.entrySet().iterator();
+				while (iterator.hasNext()) {
+					Entry<String, FileState> entry = iterator.next();
+					FileState state = entry.getValue();
+					if (state.lastWrite < expired) {
```

Comment:
Good catch but I don't think that will help See my next commit for what I think will work.